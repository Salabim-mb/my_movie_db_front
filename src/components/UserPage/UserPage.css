.userCard {
    text-align: center;
    margin-left: 6%;
}

.userBoxCard {
    margin-left: 8rem;
    margin-right: 8rem;
    margin-top: 5rem;
}

.ThumbnailImage {
    max-width: 180px;
    margin: 2%;
}

.userText {
    font-size: 30px;
}

.Button {
    display: flex;
    position: absolute;
    z-index: 10;
    right: 0;
}