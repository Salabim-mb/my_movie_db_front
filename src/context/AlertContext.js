import React, {useState, createContext} from 'react';

export const AlertContext = createContext({
    open: false,
    message: "",
    setOpen: () => {}
});

export const AlertProvider = (props) => {
    const [open, setOpen] = false;
    const [message, setMessage] = useState("");
    const [variant, setVariant] = useState("success");

    const data = {
        open,
        message,
        variant,
        changeVisibility: (newOpen) => setOpen(newOpen),
        changeMessage: (newMessage, newVariant = "success") => {
            setOpen(false);
            setMessage(newMessage);
            setVariant(newVariant);
        },
        showAlert: (newMessage, newVariant) => {
            data.changeMessage(newMessage, newVariant);
            data.changeVisibility(true);
        },
    };
    return <AlertContext.Provider value={data} {...props} />;
};