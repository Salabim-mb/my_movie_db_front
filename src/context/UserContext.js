import React, {useState, useEffect} from 'react';
import Cookies from "universal-cookie";
import {getHeaders} from "../utils/CORSHeaders";
import {backend} from "../constants/backend";

export const UserContext = React.createContext({
    token: undefined,
    data: undefined,
    login: () => {},
    logout: () => {}
});
const cookies = new Cookies();

const getUserData = async (token) => {
    const url = `${backend.USERS}data/`;
    const response = await fetch(url, {
        method: "GET",
        headers: getHeaders(token),
    });

    if (response.status === 200) {
        return await response.json();
    } else {
        throw response.status;
    }
};

export const UserProvider = (props) => {
    const [token, setToken] = useState(cookies.get("token"));
    const [data, setData] = useState(cookies.get("data"));

    useEffect(() => {
        const fetchUserData = async() => {
            try {
                const res = await getUserData(token);
                setData(res.data);
                cookies.set("data", res.data, {
                    path: "/",
                    expires: setCookieDate()
                });
            } catch(e) {
                console.log(e);
            }
        }
        if (token) {
            fetchUserData();
        }
    }, [token]);

    const user = {
        token,
        data,
        changeToken: (newToken) => {
            cookies.set("token", newToken, {
                path: "/",
                expires: setCookieDate()
            });
            setToken(newToken);
        },
        changeData: (newData) => {
            cookies.set("data", newData, {
                path: "/",
                expires: setCookieDate()
            });
            setData(newData);
        },
        login: (newToken, newData) => {
            user.changeToken(newToken);
            user.changeData(newData);
        },
        logout: () => {
            cookies.remove("token", { path: "/" });
            cookies.remove("data", { path: "/" });
            setToken(undefined);
            setData(undefined);
        }
    };

    return <UserContext.Provider value={user} {...props} />;
};

const setCookieDate = () => {
    let date = new Date();
    date.setDate(date.getDate() + 1);
    return date;
};