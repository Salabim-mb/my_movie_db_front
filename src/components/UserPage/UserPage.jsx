import React, {useContext, useEffect, useState} from "react";
import {UserContext} from "../../context/UserContext";
import "./UserPage.css"
import Card from "react-bootstrap/Card";
import {IndexLinkContainer} from "react-router-bootstrap";
import {paths} from "../../constants/routes";
import Button from "react-bootstrap/Button";
import {Alert} from "react-bootstrap";

const UserPage = () => {
    let user = useContext(UserContext);
    const [getError, setGetError] = useState(false);

    useEffect(() => {
        if (user.data === undefined) {
            setGetError(true);

        }
    }, []);

    return (getError ? (
            <Alert variant="danger">Something went wrong while trying to fetch user data.</Alert>
        ) : (
            <div className="userBoxCard">
                <div className="card mb-3">
                    <div className="row no-gutters">
                        <div className="ThumbnailImage">
                            <Card.Img alt="User icon"
                                      src="https://cdn4.iconfinder.com/data/icons/user-actions-14/24/user_movie_video_account_profile-512.png"/>
                        </div>

                        <div className="Button">
                            <IndexLinkContainer to={paths.MOVIES}>
                                <Button variant="primary" size="sm">Movies</Button>
                            </IndexLinkContainer>
                        </div>
                        <div className="col-sm-9">
                            <div className="userCard">
                                <div className="card-body">
                                    <Card.Title className="userText">
                                        {user?.data?.name}
                                    </Card.Title>

                                    <Card.Subtitle className="userText text-muted">
                                        {user?.data?.surname}
                                    </Card.Subtitle>
                                    <hr/>
                                    <Card.Text className="userText">
                                        {user?.data?.email}
                                    </Card.Text>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        )
    )
};

export default UserPage;