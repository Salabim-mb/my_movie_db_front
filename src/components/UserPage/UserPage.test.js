import React from "react";
import UserPage from "./UserPage";
import {fireEvent, waitForElement, render} from '@testing-library/react';
import {MemoryRouter, Router} from "react-router-dom";
import {createMemoryHistory} from 'history';
import {UserContext} from "../../context/UserContext";

const renderWithRouter = (
    ui,
    {
        route = "/user",
        history = createMemoryHistory({ initialEntries: [route] })
    } = {}
) => {
    return {
        ...render(
            <Router history={history}>{ui}</Router>
        ),
        history
    }
};

describe("UserPage", () => {
    let apiFail;
    let context;
    let userData;

    beforeAll(() => {
        global.fetch = jest.fn().mockImplementation((input, init) => {
            return new Promise((resolve) => {
                if (apiFail && !context?.data) {
                    resolve({status: 500})
                } else {
                    resolve({status: 200, json: () => Promise.resolve(userData)});
                }
            })
        });
    });

    beforeEach(() => {
        jest.clearAllMocks();
        apiFail = false;
        context = {
            token: "123",
            data: {
                name: "Jan",
                surname: "Kowalski",
                email: "jkowal@wp.pl"
            }
        };
        userData = {
            name: "John",
            surname: "Doe",
            email: "jdoe@example.com"
        };
    });

    it("should match snapshot", () => {
        const {container} = render(
            <UserContext.Provider value={context}>
                <MemoryRouter>
                    <UserPage />
                </MemoryRouter>
            </UserContext.Provider>
        );

        expect(container).toMatchSnapshot();
    });

    it("should render data if context is defined and api failed", async() => {
        let apiFail = true;
        const {getByText} = render(
            <UserContext.Provider value={context}>
                <MemoryRouter>
                    <UserPage />
                </MemoryRouter>
            </UserContext.Provider>
        );

        await waitForElement(() => getByText("First name:"));

        expect(getByText(""))
    });

    it("should render error on missing context and api fail", async() => {
        apiFail = true;
        context.data = undefined;
        const {getByText} = render(
            <UserContext.Provider value={context}>
                <MemoryRouter>
                    <UserPage />
                </MemoryRouter>
            </UserContext.Provider>
        );

        await expect(fetch).toHaveBeenCalledTimes(1);

        expect(getByText("Something went wrong", {exact: false})).toBeInTheDocument();
    });

    it("should fetch user data on not full context", async() => {
        const {getByText} = render(
            <UserContext.Provider value={context}>
                <MemoryRouter>
                    <UserPage />
                </MemoryRouter>
            </UserContext.Provider>
        );
    });

    it("should redirect to movie page", () => {

    });
});