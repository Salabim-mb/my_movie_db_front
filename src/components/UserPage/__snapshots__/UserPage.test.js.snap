// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPage should match snapshot 1`] = `
<div>
  <div
    class="userBoxCard"
  >
    <div
      class="card mb-3"
    >
      <div
        class="row no-gutters"
      >
        <div
          class="ThumbnailImage"
        >
          <img
            alt="User icon"
            class="card-img"
            src="https://cdn4.iconfinder.com/data/icons/user-actions-14/24/user_movie_video_account_profile-512.png"
          />
        </div>
        <div
          class="Button"
        >
          <a
            class="btn btn-primary btn-sm"
            href="/movies"
          >
            Movies
          </a>
        </div>
        <div
          class="col-sm-9"
        >
          <div
            class="userCard"
          >
            <div
              class="card-body"
            >
              <div
                class="userText card-title h5"
              >
                Jan
              </div>
              <div
                class="userText text-muted card-subtitle h6"
              >
                Kowalski
              </div>
              <hr />
              <p
                class="userText card-text"
              >
                jkowal@wp.pl
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
